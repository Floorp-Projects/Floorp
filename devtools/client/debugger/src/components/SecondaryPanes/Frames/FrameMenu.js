/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

import { showMenu } from "../../../context-menu/menu";
import { copyToTheClipboard } from "../../../utils/clipboard";
import { kebabCase } from "lodash";

const blackboxString = "ignoreContextItem.ignore";
const unblackboxString = "ignoreContextItem.unignore";

function formatMenuElement(labelString, click, disabled = false) {
  const label = L10N.getStr(labelString);
  const accesskey = L10N.getStr(`${labelString}.accesskey`);
  const id = `node-menu-${kebabCase(label)}`;
  return {
    id,
    label,
    accesskey,
    disabled,
    click,
  };
}

function copySourceElement(url) {
  return formatMenuElement("copySourceUri2", () => copyToTheClipboard(url));
}

function copyStackTraceElement(copyStackTrace) {
  return formatMenuElement("copyStackTrace", () => copyStackTrace());
}

function toggleFrameworkGroupingElement(
  toggleFrameworkGrouping,
  frameworkGroupingOn
) {
  const actionType = frameworkGroupingOn
    ? "framework.disableGrouping"
    : "framework.enableGrouping";

  return formatMenuElement(actionType, () => toggleFrameworkGrouping());
}

function blackBoxSource(cx, source, toggleBlackBox) {
  const toggleBlackBoxString = source.isBlackBoxed
    ? unblackboxString
    : blackboxString;

  return formatMenuElement(toggleBlackBoxString, () =>
    toggleBlackBox(cx, source)
  );
}

function restartFrame(cx, frame, restart) {
  return formatMenuElement("restartFrame", () => restart(cx, frame));
}

function isValidRestartFrame(frame, callbacks) {
  // Hides 'Restart Frame' item for call stack groups context menu,
  // otherwise can be misleading for the user which frame gets restarted.
  if (!callbacks.restart) {
    return false;
  }

  // Any frame state than 'on-stack' is either dismissed by the server
  // or can potentially cause unexpected errors.
  // Global frame has frame.callee equal to null and can't be restarted.
  return frame.type === "call" && frame.state === "on-stack";
}

export default function FrameMenu(
  frame,
  frameworkGroupingOn,
  callbacks,
  event,
  cx
) {
  event.stopPropagation();
  event.preventDefault();

  const menuOptions = [];

  if (isValidRestartFrame(frame, callbacks)) {
    const restartFrameItem = restartFrame(cx, frame, callbacks.restart);
    menuOptions.push(restartFrameItem);
  }

  const toggleFrameworkElement = toggleFrameworkGroupingElement(
    callbacks.toggleFrameworkGrouping,
    frameworkGroupingOn
  );
  menuOptions.push(toggleFrameworkElement);

  const { source } = frame;
  if (source) {
    const copySourceUri2 = copySourceElement(source.url);
    menuOptions.push(copySourceUri2);
    menuOptions.push(blackBoxSource(cx, source, callbacks.toggleBlackBox));
  }

  const copyStackTraceItem = copyStackTraceElement(callbacks.copyStackTrace);

  menuOptions.push(copyStackTraceItem);

  showMenu(event, menuOptions);
}
