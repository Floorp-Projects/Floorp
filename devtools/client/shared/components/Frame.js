/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this file,
 * You can obtain one at http://mozilla.org/MPL/2.0/. */

"use strict";

const { Component } = require("devtools/client/shared/vendor/react");
const dom = require("devtools/client/shared/vendor/react-dom-factories");
const PropTypes = require("devtools/client/shared/vendor/react-prop-types");
const {
  getUnicodeUrl,
  getUnicodeUrlPath,
  getUnicodeHostname,
} = require("devtools/client/shared/unicode-url");
const {
  getSourceNames,
  parseURL,
  getSourceMappedFile,
} = require("devtools/client/shared/source-utils");
const { LocalizationHelper } = require("devtools/shared/l10n");
const { MESSAGE_SOURCE } = require("devtools/client/webconsole/constants");

const l10n = new LocalizationHelper(
  "devtools/client/locales/components.properties"
);
const webl10n = new LocalizationHelper(
  "devtools/client/locales/webconsole.properties"
);

function savedFrameToLocation(frame) {
  const { source: url, line, column, sourceId } = frame;
  return {
    url,
    line,
    column,
    // The sourceId will be a string if it's a source actor ID, otherwise
    // it is either a Spidermonkey-internal ID from a SavedFrame or missing,
    // and in either case we can't use the ID for anything useful.
    id: typeof sourceId === "string" ? sourceId : null,
  };
}

class Frame extends Component {
  static get propTypes() {
    return {
      // SavedFrame, or an object containing all the required properties.
      frame: PropTypes.shape({
        functionDisplayName: PropTypes.string,
        // This could be a SavedFrame with a numeric sourceId, or it could
        // be a SavedFrame-like client-side object, in which case the
        // "sourceId" will be a source actor ID.
        sourceId: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
        source: PropTypes.string.isRequired,
        line: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
        column: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
      }).isRequired,
      // Clicking on the frame link -- probably should link to the debugger.
      onClick: PropTypes.func,
      // Option to display a function name before the source link.
      showFunctionName: PropTypes.bool,
      // Option to display a function name even if it's anonymous.
      showAnonymousFunctionName: PropTypes.bool,
      // Option to display a host name after the source link.
      showHost: PropTypes.bool,
      // Option to display a host name if the filename is empty or just '/'
      showEmptyPathAsHost: PropTypes.bool,
      // Option to display a full source instead of just the filename.
      showFullSourceUrl: PropTypes.bool,
      // Service to enable the source map feature for console.
      sourceMapURLService: PropTypes.object,
      // The source of the message
      messageSource: PropTypes.string,
    };
  }

  static get defaultProps() {
    return {
      showFunctionName: false,
      showAnonymousFunctionName: false,
      showHost: false,
      showEmptyPathAsHost: false,
      showFullSourceUrl: false,
    };
  }

  constructor(props) {
    super(props);
    this.state = {
      originalLocation: null,
    };
    this._locationChanged = this._locationChanged.bind(this);
  }

  componentWillMount() {
    if (this.props.sourceMapURLService) {
      const location = savedFrameToLocation(this.props.frame);
      // Many things that make use of this component either:
      // a) Pass in no sourceId because they have no way to know.
      // b) Pass in no sourceId because the actor wasn't created when the
      //    server sent its response.
      //
      // and due to that, we need to use subscribeByLocation in order to
      // handle both cases with an without an ID.
      this.unsubscribeSourceMapService = this.props.sourceMapURLService.subscribeByLocation(
        location,
        this._locationChanged
      );
    }
  }

  componentWillUnmount() {
    if (this.unsubscribeSourceMapService) {
      this.unsubscribeSourceMapService();
    }
  }

  _locationChanged(originalLocation) {
    this.setState({ originalLocation });
  }

  // eslint-disable-next-line complexity
  render() {
    const {
      frame,
      onClick,
      showFunctionName,
      showAnonymousFunctionName,
      showHost,
      showEmptyPathAsHost,
      showFullSourceUrl,
      messageSource,
    } = this.props;
    const { originalLocation } = this.state;

    const generatedLocation = savedFrameToLocation(frame);
    const currentLocation = originalLocation || generatedLocation;

    const source = currentLocation.url || "";
    const line =
      currentLocation.line != void 0 ? Number(currentLocation.line) : null;
    const column =
      currentLocation.column != void 0 ? Number(currentLocation.column) : null;

    const { short, long, host } = getSourceNames(source);
    const unicodeShort = getUnicodeUrlPath(short);
    const unicodeLong = getUnicodeUrl(long);
    const unicodeHost = host ? getUnicodeHostname(host) : "";

    // Reparse the URL to determine if we should link this; `getSourceNames`
    // has already cached this indirectly. We don't want to attempt to
    // link to "self-hosted" and "(unknown)".
    // Source mapped sources might not necessary linkable, but they
    // are still valid in the debugger.
    // If we have a source ID then we can show the source in the debugger.
    const isLinkable =
      originalLocation ||
      generatedLocation.id ||
      !!parseURL(generatedLocation.url);

    const elements = [];
    const sourceElements = [];
    let sourceEl;
    let tooltip = unicodeLong;

    // Exclude all falsy values, including `0`, as line numbers start with 1.
    if (line) {
      tooltip += `:${line}`;
      // Intentionally exclude 0
      if (column) {
        tooltip += `:${column}`;
      }
    }

    const attributes = {
      "data-url": long,
      className: "frame-link",
    };

    if (showFunctionName) {
      const functionDisplayName = frame.functionDisplayName;
      if (functionDisplayName || showAnonymousFunctionName) {
        elements.push(
          dom.span(
            {
              key: "function-display-name",
              className: "frame-link-function-display-name",
            },
            functionDisplayName ||
              webl10n.getStr("stacktrace.anonymousFunction")
          ),
          " "
        );
      }
    }

    let displaySource = showFullSourceUrl ? unicodeLong : unicodeShort;
    if (originalLocation) {
      displaySource = getSourceMappedFile(displaySource);
    } else if (
      showEmptyPathAsHost &&
      (displaySource === "" || displaySource === "/")
    ) {
      displaySource = host;
    }

    sourceElements.push(
      dom.span(
        {
          key: "filename",
          className: "frame-link-filename",
        },
        displaySource
      )
    );

    // If we have a line number > 0.
    if (line) {
      let lineInfo = `:${line}`;
      // Add `data-line` attribute for testing
      attributes["data-line"] = line;

      // Intentionally exclude 0
      if (column) {
        lineInfo += `:${column}`;
        // Add `data-column` attribute for testing
        attributes["data-column"] = column;
      }

      sourceElements.push(
        dom.span(
          {
            key: "line",
            className: "frame-link-line",
          },
          lineInfo
        )
      );
    }

    // Inner el is useful for achieving ellipsis on the left and correct LTR/RTL
    // ordering. See CSS styles for frame-link-source-[inner] and bug 1290056.
    let tooltipMessage;
    if (messageSource && messageSource === MESSAGE_SOURCE.CSS) {
      tooltipMessage = l10n.getFormatStr(
        "frame.viewsourceinstyleeditor",
        tooltip
      );
    } else {
      tooltipMessage = l10n.getFormatStr("frame.viewsourceindebugger", tooltip);
    }

    const sourceInnerEl = dom.span(
      {
        key: "source-inner",
        className: "frame-link-source-inner",
        title: isLinkable ? tooltipMessage : tooltip,
      },
      sourceElements
    );

    // If source is not a URL (self-hosted, eval, etc.), don't make
    // it an anchor link, as we can't link to it.
    if (isLinkable) {
      sourceEl = dom.a(
        {
          onClick: e => {
            e.preventDefault();
            e.stopPropagation();

            onClick(generatedLocation);
          },
          href: source,
          className: "frame-link-source",
          draggable: false,
        },
        sourceInnerEl
      );
    } else {
      sourceEl = dom.span(
        {
          key: "source",
          className: "frame-link-source",
        },
        sourceInnerEl
      );
    }
    elements.push(sourceEl);

    if (showHost && unicodeHost) {
      elements.push(" ");
      elements.push(
        dom.span(
          {
            key: "host",
            className: "frame-link-host",
          },
          unicodeHost
        )
      );
    }

    return dom.span(attributes, ...elements);
  }
}

module.exports = Frame;
