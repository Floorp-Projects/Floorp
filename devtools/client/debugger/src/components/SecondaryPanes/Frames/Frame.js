/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

import React, { Component, memo } from "react";
import PropTypes from "prop-types";

import classNames from "classnames";

import AccessibleImage from "../../shared/AccessibleImage";
import { formatDisplayName } from "../../../utils/pause/frames";
import { getFilename, getFileURL } from "../../../utils/source";
import FrameMenu from "./FrameMenu";
import FrameIndent from "./FrameIndent";

function FrameTitle({ frame, options = {}, l10n }) {
  const displayName = formatDisplayName(frame, options, l10n);
  return <span className="title">{displayName}</span>;
}

const FrameLocation = memo(({ frame, displayFullUrl = false }) => {
  if (!frame.source) {
    return null;
  }

  if (frame.library) {
    return (
      <span className="location">
        {frame.library}
        <AccessibleImage
          className={`annotation-logo ${frame.library.toLowerCase()}`}
        />
      </span>
    );
  }

  const { location, source } = frame;
  const filename = displayFullUrl
    ? getFileURL(source, false)
    : getFilename(source);

  return (
    <span className="location" title={source.url}>
      <span className="filename">{filename}</span>:
      <span className="line">{location.line}</span>
    </span>
  );
});

FrameLocation.displayName = "FrameLocation";

export default class FrameComponent extends Component {
  static defaultProps = {
    hideLocation: false,
    shouldMapDisplayName: true,
    disableContextMenu: false,
  };

  get isSelectable() {
    return this.props.panel == "webconsole";
  }

  get isDebugger() {
    return this.props.panel == "debugger";
  }

  onContextMenu(event) {
    const {
      frame,
      copyStackTrace,
      toggleFrameworkGrouping,
      toggleBlackBox,
      frameworkGroupingOn,
      cx,
      restart,
    } = this.props;
    FrameMenu(
      frame,
      frameworkGroupingOn,
      { copyStackTrace, toggleFrameworkGrouping, toggleBlackBox, restart },
      event,
      cx
    );
  }

  onMouseDown(e, frame, selectedFrame) {
    if (e.button !== 0) {
      return;
    }

    this.props.selectFrame(this.props.cx, frame);
  }

  onKeyUp(event, frame, selectedFrame) {
    if (event.key != "Enter") {
      return;
    }

    this.props.selectFrame(this.props.cx, frame);
  }

  render() {
    const {
      frame,
      selectedFrame,
      hideLocation,
      shouldMapDisplayName,
      displayFullUrl,
      getFrameTitle,
      disableContextMenu,
    } = this.props;
    const { l10n } = this.context;

    const className = classNames("frame", {
      selected: selectedFrame && selectedFrame.id === frame.id,
    });

    if (!frame.source) {
      throw new Error("no frame source");
    }

    const title = getFrameTitle
      ? getFrameTitle(
          `${getFileURL(frame.source, false)}:${frame.location.line}`
        )
      : undefined;

    return (
      <div
        role="listitem"
        key={frame.id}
        className={className}
        onMouseDown={e => this.onMouseDown(e, frame, selectedFrame)}
        onKeyUp={e => this.onKeyUp(e, frame, selectedFrame)}
        onContextMenu={disableContextMenu ? null : e => this.onContextMenu(e)}
        tabIndex={0}
        title={title}
      >
        {frame.asyncCause && (
          <span className="location-async-cause">
            {this.isSelectable && <FrameIndent />}
            {this.isDebugger ? (
              <span className="async-label">{frame.asyncCause}</span>
            ) : (
              l10n.getFormatStr("stacktrace.asyncStack", frame.asyncCause)
            )}
            {this.isSelectable && <br className="clipboard-only" />}
          </span>
        )}
        {this.isSelectable && <FrameIndent />}
        <FrameTitle
          frame={frame}
          options={{ shouldMapDisplayName }}
          l10n={l10n}
        />
        {!hideLocation && <span className="clipboard-only"> </span>}
        {!hideLocation && (
          <FrameLocation frame={frame} displayFullUrl={displayFullUrl} />
        )}
        {this.isSelectable && <br className="clipboard-only" />}
      </div>
    );
  }
}

FrameComponent.displayName = "Frame";
FrameComponent.contextTypes = { l10n: PropTypes.object };
