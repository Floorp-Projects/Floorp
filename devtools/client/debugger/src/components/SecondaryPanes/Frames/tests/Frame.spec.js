/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

import React from "react";
import { shallow, mount } from "enzyme";
import Frame from "../Frame.js";
import {
  makeMockFrame,
  makeMockSource,
  mockthreadcx,
} from "../../../../utils/test-mockup";

import FrameMenu from "../FrameMenu";
jest.mock("../FrameMenu", () => jest.fn());

function frameProperties(frame, selectedFrame, overrides = {}) {
  return {
    cx: mockthreadcx,
    frame,
    selectedFrame,
    copyStackTrace: jest.fn(),
    contextTypes: {},
    selectFrame: jest.fn(),
    selectLocation: jest.fn(),
    toggleBlackBox: jest.fn(),
    displayFullUrl: false,
    frameworkGroupingOn: false,
    panel: "webconsole",
    toggleFrameworkGrouping: null,
    restart: jest.fn(),
    ...overrides,
  };
}

function render(frameToSelect = {}, overrides = {}, propsOverrides = {}) {
  const source = makeMockSource("foo-view.js");
  const defaultFrame = makeMockFrame("1", source, undefined, 10, "renderFoo");

  const frame = { ...defaultFrame, ...overrides };
  const selectedFrame = { ...frame, ...frameToSelect };

  const props = frameProperties(frame, selectedFrame, propsOverrides);
  const component = shallow(<Frame {...props} />);
  return { component, props };
}

describe("Frame", () => {
  it("user frame", () => {
    const { component } = render();
    expect(component).toMatchSnapshot();
  });

  it("user frame (not selected)", () => {
    const { component } = render({ id: "2" });
    expect(component).toMatchSnapshot();
  });

  it("library frame", () => {
    const source = makeMockSource("backbone.js");
    const backboneFrame = {
      ...makeMockFrame("3", source, undefined, 12, "updateEvents"),
      library: "backbone",
    };

    const { component } = render({ id: "3" }, backboneFrame);
    expect(component).toMatchSnapshot();
  });

  it("filename only", () => {
    const source = makeMockSource(
      "https://firefox.com/assets/src/js/foo-view.js"
    );
    const frame = makeMockFrame("1", source, undefined, 10, "renderFoo");

    const props = frameProperties(frame, null);
    const component = mount(<Frame {...props} />);
    expect(component.text()).toBe("    renderFoo foo-view.js:10");
  });

  it("full URL", () => {
    const url = `https://${"a".repeat(100)}.com/assets/src/js/foo-view.js`;
    const source = makeMockSource(url);
    const frame = makeMockFrame("1", source, undefined, 10, "renderFoo");

    const props = frameProperties(frame, null, { displayFullUrl: true });
    const component = mount(<Frame {...props} />);
    expect(component.text()).toBe(`    renderFoo ${url}:10`);
  });

  it("renders asyncCause", () => {
    const url = `https://example.com/async.js`;
    const source = makeMockSource(url);
    const frame = makeMockFrame("1", source, undefined, 10, "timeoutFn");
    frame.asyncCause = "setTimeout handler";

    const props = frameProperties(frame);
    const component = mount(<Frame {...props} />, { context: { l10n: L10N } });
    expect(component.find(".location-async-cause").text()).toBe(
      `    (Async: setTimeout handler)`
    );
  });

  it("getFrameTitle", () => {
    const url = `https://${"a".repeat(100)}.com/assets/src/js/foo-view.js`;
    const source = makeMockSource(url);
    const frame = makeMockFrame("1", source, undefined, 10, "renderFoo");

    const props = frameProperties(frame, null, {
      getFrameTitle: x => `Jump to ${x}`,
    });
    const component = shallow(<Frame {...props} />);
    expect(component.prop("title")).toBe(`Jump to ${url}:10`);
    expect(component).toMatchSnapshot();
  });

  describe("mouse events", () => {
    it("does not call FrameMenu when disableContextMenu is true", () => {
      const { component } = render(undefined, undefined, {
        disableContextMenu: true,
      });

      const mockEvent = "mockEvent";
      component.simulate("contextmenu", mockEvent);

      expect(FrameMenu).toHaveBeenCalledTimes(0);
    });

    it("calls FrameMenu on right click", () => {
      const { component, props } = render();
      const {
        copyStackTrace,
        toggleFrameworkGrouping,
        toggleBlackBox,
        cx,
        restart,
      } = props;
      const mockEvent = "mockEvent";
      component.simulate("contextmenu", mockEvent);

      expect(FrameMenu).toHaveBeenCalledWith(
        props.frame,
        props.frameworkGroupingOn,
        {
          copyStackTrace,
          toggleFrameworkGrouping,
          toggleBlackBox,
          restart,
        },
        mockEvent,
        cx
      );
    });
  });
});
