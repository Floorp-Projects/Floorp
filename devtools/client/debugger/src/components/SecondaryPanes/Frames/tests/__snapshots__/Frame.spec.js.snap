// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Frame getFrameTitle 1`] = `
<div
  className="frame"
  key="1"
  onContextMenu={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  role="listitem"
  tabIndex={0}
  title="Jump to https://aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa.com/assets/src/js/foo-view.js:10"
>
  <FrameIndent />
  <FrameTitle
    frame={
      Object {
        "asyncCause": null,
        "displayName": "renderFoo",
        "generatedLocation": Object {
          "line": 10,
          "sourceId": "source",
        },
        "id": "1",
        "index": 0,
        "location": Object {
          "line": 10,
          "sourceId": "source",
        },
        "scope": Object {
          "actor": "scope-actor",
          "bindings": Object {
            "arguments": Array [],
            "variables": Object {},
          },
          "function": null,
          "object": null,
          "parent": null,
          "scopeKind": "",
          "type": "block",
        },
        "source": Object {
          "extensionName": null,
          "id": "source",
          "isBlackBoxed": false,
          "isExtension": false,
          "isOriginal": false,
          "isPrettyPrinted": false,
          "isWasm": false,
          "relativeUrl": "https://aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa.com/assets/src/js/foo-view.js",
          "url": "https://aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa.com/assets/src/js/foo-view.js",
        },
        "state": "on-stack",
        "this": Object {},
        "thread": "FakeThread",
        "type": "call",
      }
    }
    options={
      Object {
        "shouldMapDisplayName": true,
      }
    }
  />
  <span
    className="clipboard-only"
  >
     
  </span>
  <FrameLocation
    displayFullUrl={false}
    frame={
      Object {
        "asyncCause": null,
        "displayName": "renderFoo",
        "generatedLocation": Object {
          "line": 10,
          "sourceId": "source",
        },
        "id": "1",
        "index": 0,
        "location": Object {
          "line": 10,
          "sourceId": "source",
        },
        "scope": Object {
          "actor": "scope-actor",
          "bindings": Object {
            "arguments": Array [],
            "variables": Object {},
          },
          "function": null,
          "object": null,
          "parent": null,
          "scopeKind": "",
          "type": "block",
        },
        "source": Object {
          "extensionName": null,
          "id": "source",
          "isBlackBoxed": false,
          "isExtension": false,
          "isOriginal": false,
          "isPrettyPrinted": false,
          "isWasm": false,
          "relativeUrl": "https://aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa.com/assets/src/js/foo-view.js",
          "url": "https://aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa.com/assets/src/js/foo-view.js",
        },
        "state": "on-stack",
        "this": Object {},
        "thread": "FakeThread",
        "type": "call",
      }
    }
  />
  <br
    className="clipboard-only"
  />
</div>
`;

exports[`Frame library frame 1`] = `
<div
  className="frame selected"
  key="3"
  onContextMenu={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  role="listitem"
  tabIndex={0}
>
  <FrameIndent />
  <FrameTitle
    frame={
      Object {
        "asyncCause": null,
        "displayName": "updateEvents",
        "generatedLocation": Object {
          "line": 12,
          "sourceId": "source",
        },
        "id": "3",
        "index": 0,
        "library": "backbone",
        "location": Object {
          "line": 12,
          "sourceId": "source",
        },
        "scope": Object {
          "actor": "scope-actor",
          "bindings": Object {
            "arguments": Array [],
            "variables": Object {},
          },
          "function": null,
          "object": null,
          "parent": null,
          "scopeKind": "",
          "type": "block",
        },
        "source": Object {
          "extensionName": null,
          "id": "source",
          "isBlackBoxed": false,
          "isExtension": false,
          "isOriginal": false,
          "isPrettyPrinted": false,
          "isWasm": false,
          "relativeUrl": "backbone.js",
          "url": "backbone.js",
        },
        "state": "on-stack",
        "this": Object {},
        "thread": "FakeThread",
        "type": "call",
      }
    }
    options={
      Object {
        "shouldMapDisplayName": true,
      }
    }
  />
  <span
    className="clipboard-only"
  >
     
  </span>
  <FrameLocation
    displayFullUrl={false}
    frame={
      Object {
        "asyncCause": null,
        "displayName": "updateEvents",
        "generatedLocation": Object {
          "line": 12,
          "sourceId": "source",
        },
        "id": "3",
        "index": 0,
        "library": "backbone",
        "location": Object {
          "line": 12,
          "sourceId": "source",
        },
        "scope": Object {
          "actor": "scope-actor",
          "bindings": Object {
            "arguments": Array [],
            "variables": Object {},
          },
          "function": null,
          "object": null,
          "parent": null,
          "scopeKind": "",
          "type": "block",
        },
        "source": Object {
          "extensionName": null,
          "id": "source",
          "isBlackBoxed": false,
          "isExtension": false,
          "isOriginal": false,
          "isPrettyPrinted": false,
          "isWasm": false,
          "relativeUrl": "backbone.js",
          "url": "backbone.js",
        },
        "state": "on-stack",
        "this": Object {},
        "thread": "FakeThread",
        "type": "call",
      }
    }
  />
  <br
    className="clipboard-only"
  />
</div>
`;

exports[`Frame user frame (not selected) 1`] = `
<div
  className="frame"
  key="1"
  onContextMenu={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  role="listitem"
  tabIndex={0}
>
  <FrameIndent />
  <FrameTitle
    frame={
      Object {
        "asyncCause": null,
        "displayName": "renderFoo",
        "generatedLocation": Object {
          "line": 10,
          "sourceId": "source",
        },
        "id": "1",
        "index": 0,
        "location": Object {
          "line": 10,
          "sourceId": "source",
        },
        "scope": Object {
          "actor": "scope-actor",
          "bindings": Object {
            "arguments": Array [],
            "variables": Object {},
          },
          "function": null,
          "object": null,
          "parent": null,
          "scopeKind": "",
          "type": "block",
        },
        "source": Object {
          "extensionName": null,
          "id": "source",
          "isBlackBoxed": false,
          "isExtension": false,
          "isOriginal": false,
          "isPrettyPrinted": false,
          "isWasm": false,
          "relativeUrl": "foo-view.js",
          "url": "foo-view.js",
        },
        "state": "on-stack",
        "this": Object {},
        "thread": "FakeThread",
        "type": "call",
      }
    }
    options={
      Object {
        "shouldMapDisplayName": true,
      }
    }
  />
  <span
    className="clipboard-only"
  >
     
  </span>
  <FrameLocation
    displayFullUrl={false}
    frame={
      Object {
        "asyncCause": null,
        "displayName": "renderFoo",
        "generatedLocation": Object {
          "line": 10,
          "sourceId": "source",
        },
        "id": "1",
        "index": 0,
        "location": Object {
          "line": 10,
          "sourceId": "source",
        },
        "scope": Object {
          "actor": "scope-actor",
          "bindings": Object {
            "arguments": Array [],
            "variables": Object {},
          },
          "function": null,
          "object": null,
          "parent": null,
          "scopeKind": "",
          "type": "block",
        },
        "source": Object {
          "extensionName": null,
          "id": "source",
          "isBlackBoxed": false,
          "isExtension": false,
          "isOriginal": false,
          "isPrettyPrinted": false,
          "isWasm": false,
          "relativeUrl": "foo-view.js",
          "url": "foo-view.js",
        },
        "state": "on-stack",
        "this": Object {},
        "thread": "FakeThread",
        "type": "call",
      }
    }
  />
  <br
    className="clipboard-only"
  />
</div>
`;

exports[`Frame user frame 1`] = `
<div
  className="frame selected"
  key="1"
  onContextMenu={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  role="listitem"
  tabIndex={0}
>
  <FrameIndent />
  <FrameTitle
    frame={
      Object {
        "asyncCause": null,
        "displayName": "renderFoo",
        "generatedLocation": Object {
          "line": 10,
          "sourceId": "source",
        },
        "id": "1",
        "index": 0,
        "location": Object {
          "line": 10,
          "sourceId": "source",
        },
        "scope": Object {
          "actor": "scope-actor",
          "bindings": Object {
            "arguments": Array [],
            "variables": Object {},
          },
          "function": null,
          "object": null,
          "parent": null,
          "scopeKind": "",
          "type": "block",
        },
        "source": Object {
          "extensionName": null,
          "id": "source",
          "isBlackBoxed": false,
          "isExtension": false,
          "isOriginal": false,
          "isPrettyPrinted": false,
          "isWasm": false,
          "relativeUrl": "foo-view.js",
          "url": "foo-view.js",
        },
        "state": "on-stack",
        "this": Object {},
        "thread": "FakeThread",
        "type": "call",
      }
    }
    options={
      Object {
        "shouldMapDisplayName": true,
      }
    }
  />
  <span
    className="clipboard-only"
  >
     
  </span>
  <FrameLocation
    displayFullUrl={false}
    frame={
      Object {
        "asyncCause": null,
        "displayName": "renderFoo",
        "generatedLocation": Object {
          "line": 10,
          "sourceId": "source",
        },
        "id": "1",
        "index": 0,
        "location": Object {
          "line": 10,
          "sourceId": "source",
        },
        "scope": Object {
          "actor": "scope-actor",
          "bindings": Object {
            "arguments": Array [],
            "variables": Object {},
          },
          "function": null,
          "object": null,
          "parent": null,
          "scopeKind": "",
          "type": "block",
        },
        "source": Object {
          "extensionName": null,
          "id": "source",
          "isBlackBoxed": false,
          "isExtension": false,
          "isOriginal": false,
          "isPrettyPrinted": false,
          "isWasm": false,
          "relativeUrl": "foo-view.js",
          "url": "foo-view.js",
        },
        "state": "on-stack",
        "this": Object {},
        "thread": "FakeThread",
        "type": "call",
      }
    }
  />
  <br
    className="clipboard-only"
  />
</div>
`;
